x/cronos/events/bindings/src/ICA.sol

```diff
@@ -3,7 +3,32 @@ pragma solidity ^0.8.4;
 
 interface IICAModule {
     event SubmitMsgsResult(uint64 seq);
+    event DestinationCallback(
+        string module,
+        string callbackType,
+        string callbackAddress,
+        string callbackResult,
+        string callbackError,
+        string callbackExecGasLimit,
+        string callbackCommitGasLimit,
+        string packetDestPort,
+        string packetDestChannel,
+        string packetSequence
+    );
+    event SourceCallback(
+        string module,
+        string callbackType,
+        string callbackAddress,
+        string callbackResult,
+        string callbackExecGasLimit,
+        string callbackCommitGasLimit,
+        string packetSrcPort,
+        string packetSrcChannel,
+        string packetSequence
+    );
     function registerAccount(string calldata connectionID, string calldata version) external payable returns (bool);
     function queryAccount(string calldata connectionID, address addr) external view returns (string memory);
     function submitMsgs(string calldata connectionID, bytes calldata data, uint256 timeout) external payable returns (uint64);
+    function onAcknowledgementPacketCallback(uint64 seq, string calldata packetSenderAddress) external payable returns (bool);
+    function onTimeoutPacketCallback(uint64 seq, string calldata packetSenderAddress) external payable returns (bool);
```

Comment:
I am not quite certain how does this works.

Upon ICA ack and timeout, this pre-compile will be called by the ICA callback hook, how does the packet sender address get called at the end?

Reply:
I haven't come up recursive call of evm yet

Reply:
> I am not quite certain how does this works.
> 
> Upon ICA ack and timeout, this pre-compile will be called by the ICA callback hook, how does the packet sender address get called at the end?

the callback middleware will get called with the `onAck/onTimeout` methods, and get the contract address from the memo field, then out implementation will call the contract's designated methods.

Reply:
> the callback middleware will get called with the onAck/onTimeout methods

Yes, I saw this in the code, however, the `onAcknowledgementPacketCallback` and `onTimeoutPacketCallback` solidity callbak is defined in the ICA pre-compile. That's where I feel confused. Shouldn't these methods be defined in the source contract?

And I didn't saw the source contract being called for now too. Is this because it is not implemented yet?

Reply:
Yes, sorry that I misunderstand before